Subset hdf5 new tutorial

donal-at-NEON opened the pull request:
All relevant files for a new tutorial that walks people through subsetting a large NEON AOP hyperspectral HDF5 file (~652Mb) into a smaller (~50Mb) file that is easier/faster to work with.

This also includes updated image files that Colin made for the hyperspectral tutorials, as well as three new _includes files that reference the new data that I uploaded to figShare. I added these six files to each of the incoming PRs because I got a bit confused on how that all works, but once the first one is merged into NEON-Data-Skills/master they should hopefully merge without issue.

@mjones01 I'm not sure how to define the urlTitle for a new tutorial, or if we want to officially make it a part of the HDF5 series (probably not necessary, but I think it fits into the tutorials/HDF5/ directory best), but if those are things that we want to do, I can add another commit or make another PR later.

mjones01 commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,241 @@
+
+## Reading in the HDF5 file
+
+First, we will load the required library for this tutorial:
```
change "library" for "package"

donal-at-NEON replied:
Easy! done.

mjones01 commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
````diff
@@ -0,0 +1,241 @@
+```{r read-in-file}
+# set working directory to ensure R can find the file we wish to import and where
+# we want to save our files. Be sure to move the download into your working directory!
+wd="~/Desktop/Hyperspectral_Tutorial/" #This will depend on your local environment
````
Update as per wd conversation.

donal-at-NEON replied:
Yes, changed to ~/Documents/data

mjones01 commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
````diff
@@ -0,0 +1,241 @@
+
+```{r create-hdf5}
+# First, create a name for the new file
+nf <- paste0(wd, "NEON_hyperspectral_tutorial_example_subset.h5")
````
Since this is the file used in the next lessons in the series, it should be named the same as the one being read into those tutorials.

donal-at-NEON replied:
Yes, this is consistent with the other PRs. However, the other hyperspectrlal tutorial .Rmd files within this particular branch/PR have not been updated (to make the PRs unique to each tutorial). So, you might not see that this is consistent depending on how you are viewing the other files.

mjones01 commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,241 @@
+# create hdf5 file
+
+h5createFile(nf)
+h5createGroup(nf, "SJER/")
```
I need more explanation for lines 131 to 135.  How do you know you want to do this? 

Perhaps a section above that shows the structure of the existing HDF5 file and you then say, "from this we want to keep X, Y, & Z". This is then the guide for why the rest is occurring.

donal-at-NEON replied:
This is a good idea, and I will lump this comment in with a couple of others that ask for an additional section exploring and explaining the existing HDF5 format, and which parts of it (groups, spatial, and spectral subsets) we want to keep.

mjones01 replied:
Makes much more sense now.  Thanks.

mjones01 commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
````diff
@@ -0,0 +1,241 @@
+```
+
+## Adding group attributes
+One of the great things about HDF5 files is that they can contain _data_ and _attributes_ within the same group. 
````
Please use asterisk instead of underscores for emphasis.  Too often underscores are also in object and file names and this can get confusing.

donal-at-NEON replied:
Ahh yes I will use this convention from now on.

mjones01 commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,241 @@
+
+## Adding group attributes
+One of the great things about HDF5 files is that they can contain _data_ and _attributes_ within the same group. 
+As explained within the <a href="{{ site.baseurl }}/intro-hdf5-r-series" target="_blank"> Introduction to Working with HDF5 Format in R series</a>, attributes are a type of metadata that are associated with an HDF5 group or dataset. There may be multiple attributes associated with each group and/or dataset. Attributes come with a name and an associated array of information. In this tutorial, we w ...
```
Add line break after hyperlink. Very long line.

donal-at-NEON replied:
I have observed the line break behavior carefully, and I think it is actually better to leave paragraphs as long lines. Once the resulting .md file is converted to a web page, somehow it is all reformatted to be continuous paragraphs of the appropriate width, regardless of how the lines are broken in the .Rmd. Moreover, sometimes there is not a space character between the last word of a line and the first word of the next line, and I have seen some tutorials where this results in the two words displaying together (without a space) somewhere in the resulting paragraph. Let's give this a try and see how it looks in the end.

mjones01 commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
````diff
@@ -0,0 +1,241 @@
+
+```{r ref-attributes}
+
+a=h5readAttributes(f,"/SJER/Reflectance/")
````
Please change all "=" assignment operators to "<-" assignment operators throughout.

donal-at-NEON replied:
Yup! Done.

mjones01 commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
````diff
@@ -0,0 +1,241 @@
+h5closeAll()
+```
+
+Next, we will loop through each of the datasets within the Coordinate_System group, and copy those (and their attributes, if present) from the full tile to our subset file. The Coordinate_System group contains many important pieces of information for geolocating our data, so we need to make sure that the sunset file has those data.
````
break line at 80 char.

donal-at-NEON replied:
Please see comment above re: automative paragraph formatting.

mjones01 commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,241 @@
+# make a list of all groups within the full tile file
+ls=h5ls(f,all=T)
+
+#make a list of all of the names within the Coordinate_System group
```
add space after # for comments in code (but not commented out code)

donal-at-NEON replied:
Good eye. I have corrected throughout

mjones01 commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,241 @@
+
+## Spectral Subsetting
+
+The goal of subsetting this dataset is to substantially reduce the file size to make it faster to download and process these data. While each AOP mosaic tile is not particularly large in terms of its spatial scale (1km by 1km at 1m resolution= 1,000,000 pixels, or about half as many pixels at shown on a standard 1080p computer screen), the 426 spectral bands available result in a fairly large fil ...
```
1) add line breaks. 
2) typo "demonstartion"
3) if it make sense, consider adding a sentence or paragraph on occasions when spectral subsetting woudn't make sense.

donal-at-NEON replied:
1. See above re: line breaks.
2. Typo corrected - good eye!
3. Not sure this is the most eloquent way to put it, but I added the information that you suggested and split this into two smaller paragraphs:
"The goal of subsetting this dataset is to substantially reduce the file size, making it faster to download and process these data. While each AOP mosaic tile is not particularly large in terms of its spatial scale (1km by 1km at 1m resolution= 1,000,000 pixels, or about half as many pixels at shown on a standard 1080p computer screen), the 426 spectral bands available result in a fairly large file size. Therefore, we will reduce the spectral resolution of these data by selecting every fourth band in the dataset, which reduces the file size to 1/4 of the original! 

Some circumstances demand the full spectral resolution file. For example, if you wanted to discern between the spectral signatures of similar minerals, or if you were conducting an analysis of the differences in the 'red edge' between plant functional types, you would want to use the full spectral resolution of the original hyperspectral dataset. Still, we can use far fewer bands for demonstration and teaching purposes, while still getting a good sense of what these hyperspectral data can do."

mjones01 commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,241 @@
+idx <- seq(from = 1, to = 426, by = 4)
+
+# We then use this index to select particular wavelengths from the full tile using the "index=" argument
+ws <- h5read(file = f, 
```
why "ws" - not clear from the comment.

donal-at-NEON replied:
shorthand for 'wavelengths', but I expanded to the full word in this revision to make it easier to read

mjones01 commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,241 @@
+
+## Spatial Subsetting
+
+Even after spectral subsetting, our file size would be greater than 100Mb. Therefore, we will perform a spatial subsetting process as well when we read the hyperspectral imagery data. Note that the `index=` argument in `h5read` requires a `list()` object of three dimenstions for this example - in the order of band, column, row. Here, we are taking every fourth band (using our `idx` variabe), colu ...
```
1) add line breaks
2) add () to `h5read` to indicate it is a function.
3) in this case, I don't believe `list()` should be referred to as an "object" - isn't it a function here that contains the elements of interest?  
4) In my earlier suggestion about adding a section exploring the HDF5 file could also add in this description of the how you get 1:500 and 501:1000?

donal-at-NEON replied:
1. Please see comment above re: line breaks.
2. Yes, should add `()` and made this correction.
3. I was wrong here, `list` should not have parentheses following. I believe that it is still technically an object (in computer programming nomenclature), but 'list' is pretty self explanatory, so I removed the word 'object'
4. Yes, that's a good idea. I will leave this 'conversation' open until I add that section and clarify these points.

mjones01 commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
````diff
@@ -0,0 +1,241 @@
+        write.attributes=T)
+```
+
+## Spatial Subsetting
````
Do you think spatial or spectral subsetting is more common?  I'd put the more common one first in the lesson if it isn't already.

donal-at-NEON replied:
I think that this is a good idea in general. Spatial subsetting is far more common than spectral subsetting, as it is often done with a single band, or standard multispectral images (e.g., Landsat). However for this particular application I think it makes sense to do the spectral subsetting first because: 1) we combine the spatial and spectral subsetting in the `h5read()` function using the `index=` argument. 2) the spatial subsetting is fairly simple, but the spectral subsetting is a bit more complicated, therefore 3) I think it makes more sense to thoroughly explain the spectral subsetting, and define the 'Wavelength' group within the HDF5 file, first, so that the `idx` variable can be passed directly to the `h5read()` function without combining too many concepts into a single line.

Still, I could also re-arrange things so that we explore/define the spatial subset earlier, and make the 1:500 and 501:1000 their own variables (vectors i.e., row_sub and col_sub). I'm going to ponder these options and leave this 'conversation' open until we find the best option.

mjones01 replied:
Solid reasoning.  I'll defer to you.

mjones01 commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,241 @@
+If you find that downloading the full `652 Mb` file takes too much
+time or storage space, you will find a link to this subsetted file
+at the top of each lesson in the <a href="{{ site.baseurl }}/intro-hsi-r-series" target="_blank"> *Introduction to Hyperspectral Remote Sensing Data* series</a>.
+
```
Two places below, I mention a section needed to help explore the data and set out a path for where the tutorial is headed so that the user understand why they are doing something and the goals.  I'm envisioning something like the following: 

## What to subset
(might move a bit of the above here).  

We're working with NEON data and we want a smaller file for data processing as we don't need all 426 bands and our study location is only in the NE corner of this tile.  Therefore, we're going to subset in two ways, by spectra and by spatial extent.  

### NEON HDF5 Format 
In order to subset, we need to figure out the elements of the file...

donal-at-NEON replied:
Yeah! I think this will really improve this tutorial by explaining these steps. I will leave this 'conversation' open until these changed are made.

mjones01 replied:
Perfect!

mjones01 commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
````diff
@@ -0,0 +1,241 @@
+
+
+```{r read-hyp-data}
+# Read in reflectance data - note the index argument! This is ultimately how we reduce the file size.
````
1) Do you mean "index" argument, or the full list() one? 
2) Why are you doing this on "f" and not building off the spectral subsetting? I can see wanting to demonstrate both and not having them dependent on each other but describing this in the text would be good.

donal-at-NEON replied:
1) ahh yes, I am trying to draw the reader's attention to the list that we are passing to the index= argument. I have clarified this in the latest commit.
2) Not sure I understand this comment. What we are doing here is extracting a subset of data from the 'f' file, saving that to a variable, adding attributes to that variable, then writing that variable to the new file 'nf' in the /Reflectance/Reflectance_Data group. In this way, the 'nf' is dependent upon the subset of the 'f' file as an input. I will leave this 'conversation' open pending your confirmation or further dispute.

donal-at-NEON replied:
'f' in this tutorial is the full NEON tile, but 'f' in the hyperspectral series is the example subset. Would be better to change the name for this 'subset' tutorial so that the output is called 'f' to be consistent with the hyperspectral tutorial series.

donal-at-NEON replied:
These changes have been completed in the latest commit

mjones01 commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,241 @@
+             name = "SJER/Reflectance/Reflectance_Data", 
+             index = list(idx, 501:1000, 1:500)
+            )
+# grab the '$dim' attribute - as this will be needed when writing the file at the bottom
```
for ease of reading, add lines between code and the next comment

donal-at-NEON replied:
Yes, it looks better formatted with a line in between! Corrected.

mjones01 commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,241 @@
+# We also need the attributes for the reflectance data, of course.
+ha <- h5readAttributes(file = f, 
+                       name = "SJER/Reflectance/Reflectance_Data")
+
```
Seems like this could use a more full explanation in text not just in comments. And then the code in a new code chunk.

donal-at-NEON replied:
Good point - I have added a paragraph and separated this part of the code into two chunks.

mjones01 commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
````diff
@@ -0,0 +1,241 @@
+h5closeAll()
+```
+
+That's it! We just created a subset of the original HDF5 file, and included the most essential groups and metadata for exploratory analysis. You may consider adding other information, such as the weather quality indicator, when subsetting datasets for your own purposes.
````
Add line break

donal-at-NEON replied:
Please see above comment re: line breaks.

donal-at-NEON commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,241 @@
+title: "Subsetting HDF5 files to make a smaller deliverable"
+code1: HDF5/Subset-HDF5-hyperspectral.R
+contributors: null
+dataProudct: null
```
Added the hyperspectral and bundled eddy covariance dpIDs

donal-at-NEON commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,241 @@
+---
+title: "Subsetting HDF5 files to make a smaller deliverable"
```
I suppose that you could use these methods to subset any HDF5 file, but this is specifically an example about NEON hyperspectral, so I changed the title per your suggestion.

donal-at-NEON commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,241 @@
+---
+title: "Subsetting HDF5 files to make a smaller deliverable"
```
Yup! done.

donal-at-NEON commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,241 @@
+code1: HDF5/Subset-HDF5-hyperspectral.R
+contributors: null
+dataProudct: null
+dateCreated: 2020-02-04 14:20:52
```
Ahh ok. Removed the time leaving date only

donal-at-NEON commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,241 @@
+contributors: null
+dataProudct: null
+dateCreated: 2020-02-04 14:20:52
+description: Take a large HDF5 file and extract only the information that you need,
```
1. That's an easy fix. 2. Specified that this is for NEON hyperspectral files.

donal-at-NEON commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,241 @@
+description: Take a large HDF5 file and extract only the information that you need,
+  then save as a new HDF5 file. For an example, we will take an existing hyperspectral
+  dataset (~600Mb) and reduce it in size for subsequent tutorials.
+estimatedTime: 1.0 Hours
```
Whoops! Fixed.

donal-at-NEON commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,241 @@
+authors: Donal O'Leary
+topics: hyperspectral, HDF5, remote-sensing
+tutorialSeries: null
+urlTitle: subset-hdf5-r
```
added 'hyperspectral'

donal-at-NEON commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,241 @@
+
+In this tutorial, we will subset an existing HDF5 file containing NEON
+hyperspectral data. The purpose of this exercise is to generate a smaller
+file for use in subsequent tuorials.
```
Changed to: "The purpose of this exercise is to generate a smaller file for use in subsequent analyses to reduce the file transfer time and processing power needed."

donal-at-NEON commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,241 @@
+
+After completing this activity, you will be able to:
+
+* Navigate an HDF5 file to identify the variables of interest.</li>
```
ok! removed.

donal-at-NEON commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,241 @@
+
+{% include/dataSubsets/_data_Imaging-Spec-Data-Full-Tile-H5-2020.html %}
+
+This file is NEON_D17_SJER_DP3_257000_4112000_reflectance.h5 from March 2019, available on data.neonscience.org.
```
That's a good idea - I removed the description that you flagged here and added this information to the _includes file.

donal-at-NEON commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,241 @@
+
+If you haven't already, we highly recommend you work through the 
+<a href="{{ site.baseurl }}/intro-hdf5-r-series" target="_blank"> *Introduction to Working with HDF5 Format in R* series</a>
+before moving on to this tutorial.
```
Good point. I added some context here, so that now it reads:
"For this tutorial, we recommend that you have some basic familiarity with the HDF5 file format, including knowing how to open HDF5 files (in Rstudio or HDF5Viewer) and how groups and metadata are structured. To brush up on these skills, we suggest that you work through the <a href="{{ site.baseurl }}/intro-hdf5-r-series" target="_blank"> *Introduction to Working with HDF5 Format in R* series</a> before moving on to this tutorial."

donal-at-NEON commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,241 @@
+
+
+## Why subset your dataset? 
+There are many reasons why you may wish to subset you HDF5 file.
```
Whoops! fixed.

donal-at-NEON commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,241 @@
+and shortening file transfer/download times. In this example, we
+will take a full HDF5 file of NEON hyperspectral reflectance data
+from the San Joaquin Experimental Range (SJER) that has a file size
+of ~`652 Mb` and make a new HDF5 file with a reduced spatial extent,
```
I was thinking that it made sense but I see how it is not consistent with the convention in other tutorials, so I have removed the code formatting.

donal-at-NEON commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,241 @@
+Primarily, HDF5 files may contain a large amount of information
+that is not necessary for your purposes. By subsetting the file,
+you can reduce file size, thereby shrinking your storage needs,
+and shortening file transfer/download times. In this example, we
```
Yes! added comments regarding processing time throughout this paragraph.

donal-at-NEON commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,241 @@
+syncID: 884cc5915a3842c1ab6617b060797fab
+authors: Donal O'Leary
+topics: hyperspectral, HDF5, remote-sensing
+tutorialSeries: null
```
I'm not sure about this one. I have considered making it the last lesson in the HDF5 series, or the first in the hyperspectral series. I think I prefer making it the last lesson in the HDF5 series, because it seems to be more of an HDF5 skill than a hyperspectral skill to me. I could also see it making sense in the context of the hyperspectral tutorial, because it does affect all of the other tutorials because we are using a slightly different dataset (in particular, with reduced spectral resolution). But I don't have a strong opinion either way, really. I will leave this unchanged for now, pending a consensus opinion.

mjones01 replied:
I'm happy to defer to you:  last lesson in the HDF5 series.  Also, we could list is as an "optional" starter tutorial on the landing page for the hyperpectral series but not include it in the listed tutorials.

donal-at-NEON replied:
I think this is a great idea! I will add a note to the Trello card for the Hyperspectral tutorial series

donal-at-NEON commented:
Commit "772b33e" is my first pass at addressing these concerns. There are still several open conversations to be addressed at a later time (probably tomorrow).
Thanks for the thorough review!

mjones01 commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,322 @@
+---
+syncID: 884cc5915a3842c1ab6617b060797fab
+title: "Subsetting NEON HDF5 hyperspectral files to make a smaller deliverable"
```
Is "deliverable" a common enough term?  What about "to reduce file size"?

donal-at-NEON replied:
Good suggestion. I have made this correction.

mjones01 commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,322 @@
+
+## Exploring the NEON hyperspectral HDF5 file structure
+
+In order to determine what information that we want to put into our subset, we should first take a look at the full NEON hyperspectral HDF5 file structure to see what is included. To do so, we will load the required package for this tutorial (you can un-comment the middle two lines to load 'BiocManager' and 'rhdf5' if you don't already have it on your computer).
```
Throughout you have many lines >80 characters.  Can you break them at 80 chars so on IDEs that don't automatically wrap text they won't run off the screen?

donal-at-NEON replied:
Ahhhhh okay now I understand the reason for doing this - IDEs that don't do the wrapping automatically! I'm sure that you had told me this before but I forgot. Thanks for the reminder. I have updated this throughout the tutorial.

mjones01 commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,322 @@
+
+## Spatial Subsetting
+
+Even after spectral subsetting, our file size would be greater than 100Mb. Therefore, we will also perform a *spatial* subsetting processto further reduce our file size. Now, we need to figure out which part of the full image that we want to extract for our subset. It takes a number of steps in order to read in a band of data and plot the reflectance values - all of which are thoroughly described ...
```
2nd line - "processto" needs a space.

donal-at-NEON replied:
Good eye! Corrected.

mjones01 commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,322 @@
+As we can see here, this hyperspectral reflectance tile contains a school campus that is under construction. There are many different land cover types contained here, which makes it a great example! Perhaps the most unique feature shown is in the bottom right corner of this image, where we can see the tip of a small reservoir. Let's be sure to capture this feature in our spatial subset, as well a ...
+
+While raster images count their pixels from the top left corner, we are working with a matrix, which counts its pixels from the bottom left corner. Therefore, rows are counted from the bottom to the top, and columns are counted from the left to the right. If we want to sample the bottom right quadrant of this image, we need to select rows 1 through 500 (bottom half), and columns 501 through 1000  ...
+
```
I have a much better understanding for why & how these numbers come about.  Thanks for the added details.

donal-at-NEON replied:
Thank _you_ for the good suggestion!

mjones01 commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,322 @@
+
+</div>
+
+## Extracting the example sbset
```
typo

mjones01 replied:
Also, is "example subset" the most useful header? Could it be just be "Extracting a subset"?

donal-at-NEON replied:
Yes this is a good suggestion, and has been incorporated in the latest version.

mjones01 commented on tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd:
```diff
@@ -0,0 +1,322 @@
+# However, some of the attributes are not longer valid since 
+# we changed the spatial extend of this dataset. therefore, 
+# we will need to overwrite those with the correct values.
+ha$Dimensions=c(500,500,107) # Note that the HDF5 file saves dimensions in a different order than R reads them
```
should the = be changed to <-  assignment operator?

donal-at-NEON replied:
Hah yes, this is going to take me a while to get used to using the <- operator..

donal-at-NEON commented:
After addressing the file variable name re: comment above, this will be ready to merge.

donal-at-NEON commented:
Merging this new tutorial after review by @mjones01
